File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) return 0;

        Stack<Pair<TreeNode, Integer>> myStack = new Stack<>();
        
        myStack.push(new Pair<TreeNode, Integer>(root, 1));

        int maxD = 0;

        while (!myStack.isEmpty())
        {
            Pair<TreeNode, Integer> top = myStack.pop();

            maxD = Math.max(maxD, top.getValue());

            if (top.getKey().right != null) {
                myStack.push(new Pair<TreeNode, Integer>(top.getKey().right, top.getValue() + 1));
            }
            if (top.getKey().left != null) {
                myStack.push(new Pair<TreeNode, Integer>(top.getKey().left, top.getValue() + 1));
            }
        }

        return maxD;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-1.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        top = 0
        bottom = len(matrix) - 1

        while top <= bottom:
            row = top + ((bottom - top) // 2)

            if target < matrix[row][0]:
                bottom = row - 1
            elif target > matrix[row][-1]:
                top = row + 1
            else:
                break

        left = 0
        right = len(matrix[row]) - 1

        while left <= right:
            m = left + ((right - left) // 2)

            if target < matrix[row][m]:
                right = m - 1
            elif target > matrix[row][m]:
                left = m + 1
            else:
                return True
                
        return False